{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "d2ca795d-c08d-46d1-b7f2-9481a9785a6b",
   "metadata": {},
   "source": [
    "### Outline\n",
    "The goal of this code is to generate recombineering primers to tag the native copy of Msm genes with fkbp-eGFP"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c8bf4149-59f2-47e8-a2b8-c15e2859ca9b",
   "metadata": {},
   "source": [
    "### Import packages and define functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "4a8a180b-11d4-451d-ad75-e3e2513ac7aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd \n",
    "import numpy as np \n",
    "from Bio.Seq import Seq \n",
    "from Bio.SeqUtils import MeltingTemp as mt\n",
    "from Bio import SeqIO\n",
    "from Bio.SeqUtils import GC\n",
    "\n",
    "\n",
    "def PCR_handle(gene_sequence, gibson_handle, direction):\n",
    "        \n",
    "    #Compute reverse complement for reverse primers\n",
    "    if direction == 'rvs':\n",
    "        gene_sequence = gene_sequence.reverse_complement()\n",
    "        \n",
    "    #Grabbing the first nucleotide to initialize a seq object\n",
    "    pcr_handle = gene_sequence[0]\n",
    "    \n",
    "    #Extend PCR handle based on gene sequence until a melting temperature of 58 reached\n",
    "    i = 1\n",
    "    while mt.Tm_NN(pcr_handle, nn_table=mt.DNA_NN4)<58:\n",
    "        pcr_handle = pcr_handle + gene_sequence[i]\n",
    "        i = i+1\n",
    "        \n",
    "    #Set the minumum length of the PCR handle to 19 nucleotides\n",
    "    if i < 19:\n",
    "        pcr_handle = pcr_handle + gene_sequence[i:19]\n",
    "        i = 19\n",
    "        \n",
    "    #Returns primer sequence and melting temp\n",
    "    primer = gibson_handle + pcr_handle\n",
    "    return(str(primer))\n",
    "\n",
    "def determine_recombineering_primers(locus_list):\n",
    "    \n",
    "    # Vector handles to insert into pHW152 digested with NdeI and NotI \n",
    "    # FKBP handles for adding the fkbp-egfp-Zeo fragment between the upstream and downstream regions \n",
    "    fwd_vector = Seq('tttaagaaggagatatacatatg').lower()\n",
    "    rv_linker6 = Seq('ACCGCTGCCCGAGCCACC').lower()\n",
    "\n",
    "    fwd_linker6 = Seq('ATctcgagtctagaagta').lower()\n",
    "    rv_vector = Seq('ggtggtggtgctcgagtgcggccgc').lower()\n",
    "    upstream=500\n",
    "    downstream=500\n",
    "    \n",
    "    gene_sequences = []\n",
    "    \n",
    "    for locus_name in locus_list:\n",
    "    \n",
    "        if locus_name not in msm_genes['Locus'].values:\n",
    "            print(f\"Locus name '{locus_name}' not found in the DataFrame.\")\n",
    "            return None\n",
    "\n",
    "        start_position = msm_genes.loc[msm_genes['Locus'] == locus_name, 'Start'].iloc[0]\n",
    "        stop_position = msm_genes.loc[msm_genes['Locus'] == locus_name, 'Stop'].iloc[0]\n",
    "        strand = msm_genes.loc[msm_genes['Locus'] == locus_name, 'Strand'].iloc[0]\n",
    "\n",
    "        start_position -= 1\n",
    "\n",
    "        if strand == '+':\n",
    "            gene_seq = msm_genome[start_position:stop_position]\n",
    "            protein = gene_seq.translate()\n",
    "            if protein.endswith('*'):\n",
    "                upstream_seq = msm_genome[stop_position - upstream:stop_position - 3]\n",
    "            else:\n",
    "                upstream_seq = msm_genome[stop_position - upstream:stop_position]\n",
    "            downstream_seq = msm_genome[stop_position:stop_position + downstream]\n",
    "\n",
    "        elif strand == '-':\n",
    "            gene_seq = msm_genome[start_position:stop_position].reverse_complement()\n",
    "            protein = gene_seq.translate()\n",
    "            if protein.endswith('*'):\n",
    "                upstream_seq = msm_genome[start_position + 3:start_position + 3 + upstream].reverse_complement()\n",
    "            else:\n",
    "                upstream_seq = msm_genome[start_position:start_position + upstream].reverse_complement()\n",
    "            downstream_seq = msm_genome[start_position - downstream:start_position].reverse_complement()\n",
    "\n",
    "        #Determine the PCR handles based on these upstream and downstream sequences \n",
    "        p1 = PCR_handle(upstream_seq, fwd_vector, 'fwd')\n",
    "        p2 = PCR_handle(upstream_seq, rv_linker6, 'rvs')\n",
    "        p3 = PCR_handle(downstream_seq, fwd_linker6, 'fwd')\n",
    "        p4 = PCR_handle(downstream_seq, rv_vector, 'rvs')\n",
    "        gene_sequences.append([locus_name, str(upstream_seq), str(downstream_seq), p1, p2, p3, p4, max(len(p1), len(p2), len(p3), len(p4))])\n",
    "\n",
    "    \n",
    "    columns = ['gene', \n",
    "               'Msm Upstream Recombineering Sequence', \n",
    "               'Msm Downstream Recombineering Sequence', \n",
    "               'Msm Fwd Upstream primer', \n",
    "               'Msm Rvs Upstream primer', \n",
    "               'Msm Fwd Downstream primer', \n",
    "               'Msm Rvs Downstream primer', \n",
    "               'Msm max recombineering primer length']\n",
    "    return pd.DataFrame(gene_sequences, columns=columns)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "71a6ce73-127a-40b8-a974-8b0e76b375fe",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Load reference sequence information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "0678846a-b373-48e0-8822-037e45fce260",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Load Msmeg gene information as a dataframe\n",
    "msm_genes = pd.read_csv(\"./Important_References/Mycobacterium_smegmatis_MC2-155_txt_v4.txt\", sep=\"\\t\")\n",
    "\n",
    "#Load Msmeg genome into a SeqRecord\n",
    "msm_genome = next(SeqIO.parse('./Important_References/Mycobacterium_smegmatis_MC2-155_genome_v4.fasta', \"fasta\")).seq\n",
    "\n",
    "#Load list of target genes for TPD \n",
    "union_ortho_table_filtered = pd.read_csv('./Important_References/union_ortho_table_filtered.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "591ad3fa-6fe1-493d-b698-59bb11ed9126",
   "metadata": {},
   "source": [
    "### Define genes and determine primers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "ea8a2b09-565d-4859-ad47-6352ef2cac3b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Define list of genes for recombineering \n",
    "target_gene_list = union_ortho_table_filtered['gene']\n",
    "\n",
    "#Run function to determine recombineering primers\n",
    "TPD_recombineering_primers = determine_primers(target_gene_list)\n",
    "\n",
    "#Add recombineering primers to union_ortho_table_random\n",
    "merged_df = pd.merge(union_ortho_table_filtered, TPD_recombineering_primers, on='gene')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a41dcab7-0e07-4ae0-a1cb-fa2bdb44626d",
   "metadata": {},
   "source": [
    "### Export data as csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "2fa6ac03-ab1b-4a5d-9eba-5ccea7f6b23b",
   "metadata": {},
   "outputs": [],
   "source": [
    "path = './Important_References/'\n",
    "merged_df.to_csv(path + 'TPD_recombineering_primers.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "5bd8207a-48c7-4985-b3f4-0584a4e54210",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>gene</th>\n",
       "      <th>rv</th>\n",
       "      <th>Msm_VI</th>\n",
       "      <th>Msm_Ess_TnSeq</th>\n",
       "      <th>Msm_Ess_CRISPR</th>\n",
       "      <th>Mtb_VI</th>\n",
       "      <th>Mtb_Ess_TnSeq</th>\n",
       "      <th>Mtb_Ess_CRISPR</th>\n",
       "      <th>Msm_seq</th>\n",
       "      <th>...</th>\n",
       "      <th>Mtb_Aliphatic_Index</th>\n",
       "      <th>PI_difference</th>\n",
       "      <th>Aliphatic_Index_difference</th>\n",
       "      <th>Msm Upstream Recombineering Sequence</th>\n",
       "      <th>Msm Downstream Recombineering Sequence</th>\n",
       "      <th>Msm Fwd Upstream primer</th>\n",
       "      <th>Msm Rvs Upstream primer</th>\n",
       "      <th>Msm Fwd Downstream primer</th>\n",
       "      <th>Msm Rvs Downstream primer</th>\n",
       "      <th>Msm max recombineering primer length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>MSMEG_0250</td>\n",
       "      <td>Rv0206c</td>\n",
       "      <td>-7.702</td>\n",
       "      <td>Essential</td>\n",
       "      <td>Essential</td>\n",
       "      <td>-11.013</td>\n",
       "      <td>Essential</td>\n",
       "      <td>Essential</td>\n",
       "      <td>MFAWWGRTVYQFRYIVIGVMVALCLGGGVYGISLGNHVTQSGFYDE...</td>\n",
       "      <td>...</td>\n",
       "      <td>102.658898</td>\n",
       "      <td>3.843500</td>\n",
       "      <td>8.927408</td>\n",
       "      <td>GCAACGCCGTTCGCAATGCGGTCAACAGCGCGGTGCACGGCGGCGC...</td>\n",
       "      <td>GGCCCGGCTACTCCCGTTCGGGAAGGTCGTCGGACTCTGCGGCGAT...</td>\n",
       "      <td>tttaagaaggagatatacatatgGCAACGCCGTTCGCAATGC</td>\n",
       "      <td>accgctgcccgagccaccCAGCCTGCCTTCGCGGCGC</td>\n",
       "      <td>atctcgagtctagaagtaGGCCCGGCTACTCCCGTTC</td>\n",
       "      <td>ggtggtggtgctcgagtgcggccgcCGTGCGCGCGACGTTCCTG</td>\n",
       "      <td>44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>MSMEG_0311</td>\n",
       "      <td>Rv0225</td>\n",
       "      <td>-0.060</td>\n",
       "      <td>Essential</td>\n",
       "      <td>Essential</td>\n",
       "      <td>-8.756</td>\n",
       "      <td>Essential</td>\n",
       "      <td>Essential</td>\n",
       "      <td>MSARPESAPHVRRVLLLCWRDTGHPQGGGSEAYVQRIGAYLAGRGV...</td>\n",
       "      <td>...</td>\n",
       "      <td>105.598958</td>\n",
       "      <td>0.352306</td>\n",
       "      <td>6.928882</td>\n",
       "      <td>AAGCCGTCGCGCAACTGCGCACCGAGATGCCTGATGTGCACCTGGA...</td>\n",
       "      <td>CGCCGCTCAGAACGCCTCGGCGGCTGCCTTCGCGGTGCGGGCCAGC...</td>\n",
       "      <td>tttaagaaggagatatacatatgAAGCCGTCGCGCAACTGCG</td>\n",
       "      <td>accgctgcccgagccaccGACCAGGCCGCTGAGGTAT</td>\n",
       "      <td>atctcgagtctagaagtaCGCCGCTCAGAACGCCTCG</td>\n",
       "      <td>ggtggtggtgctcgagtgcggccgcCCGCGGCGCTCGACGCCAT</td>\n",
       "      <td>44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>MSMEG_0317</td>\n",
       "      <td>Rv0227c</td>\n",
       "      <td>-9.654</td>\n",
       "      <td>Essential</td>\n",
       "      <td>Essential</td>\n",
       "      <td>-13.167</td>\n",
       "      <td>Essential</td>\n",
       "      <td>Essential</td>\n",
       "      <td>MNRAVALRIAACGLLGLGAALLIAALLLTTYTKGKIAKIPLDIDTS...</td>\n",
       "      <td>...</td>\n",
       "      <td>76.294537</td>\n",
       "      <td>0.287468</td>\n",
       "      <td>9.785865</td>\n",
       "      <td>TCACCGCGCGCGCCGAGGTGTGGGGTGTGCCGGGCGAACCGGACGA...</td>\n",
       "      <td>GAACTGCGGTCCTGCTGCGGTCATGCTCGGTGCCGGCCTATGCGCT...</td>\n",
       "      <td>tttaagaaggagatatacatatgTCACCGCGCGCGCCGAGGT</td>\n",
       "      <td>accgctgcccgagccaccGATCGGTCGGTCCGGTGGC</td>\n",
       "      <td>atctcgagtctagaagtaGAACTGCGGTCCTGCTGCG</td>\n",
       "      <td>ggtggtggtgctcgagtgcggccgcCATCCAGAACACCAGCGGAGC</td>\n",
       "      <td>46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>MSMEG_0359</td>\n",
       "      <td>Rv0236c</td>\n",
       "      <td>-0.614</td>\n",
       "      <td>Essential</td>\n",
       "      <td>Essential</td>\n",
       "      <td>-6.080</td>\n",
       "      <td>Essential</td>\n",
       "      <td>Essential</td>\n",
       "      <td>MVAAATLVLTFAQSPGQISPDTKLDLTANPLRFLARAFNLWNSDLP...</td>\n",
       "      <td>...</td>\n",
       "      <td>100.614286</td>\n",
       "      <td>3.159969</td>\n",
       "      <td>2.127723</td>\n",
       "      <td>ACATGCCGTACCGCGTCGGACTGATCGGCGGACTGGCACTCCTGCC...</td>\n",
       "      <td>CCAATCCTTGCCACCTCAACTGGTTGACAGCGAGCCGGTGCAGTAC...</td>\n",
       "      <td>tttaagaaggagatatacatatgACATGCCGTACCGCGTCGG</td>\n",
       "      <td>accgctgcccgagccaccACCGGCGTGAGCGGATGCT</td>\n",
       "      <td>atctcgagtctagaagtaCCAATCCTTGCCACCTCAACTG</td>\n",
       "      <td>ggtggtggtgctcgagtgcggccgcCGCGCTGTACGGCGACGAC</td>\n",
       "      <td>44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>MSMEG_0384</td>\n",
       "      <td>Rv0334</td>\n",
       "      <td>-9.876</td>\n",
       "      <td>Essential</td>\n",
       "      <td>Essential</td>\n",
       "      <td>-2.498</td>\n",
       "      <td>Essential</td>\n",
       "      <td>Essential</td>\n",
       "      <td>MRGIILAGGSGTRLHPLTIGVSKQLLPVYDKPLVYYPLSTLIMAGI...</td>\n",
       "      <td>...</td>\n",
       "      <td>99.930556</td>\n",
       "      <td>0.132063</td>\n",
       "      <td>1.388889</td>\n",
       "      <td>GCTTCGAACACGTCAGTGGTGGGGCAATATTCGCGTACTGGGTGGC...</td>\n",
       "      <td>CGTGCCTAGCGCAAGGGCATACTGCTGTAACACCTGGCATGCCCAC...</td>\n",
       "      <td>tttaagaaggagatatacatatgGCTTCGAACACGTCAGTGGTGG</td>\n",
       "      <td>accgctgcccgagccaccCTCTCGATCCAGAAGTTGGAGCAGG</td>\n",
       "      <td>atctcgagtctagaagtaCGTGCCTAGCGCAAGGGCA</td>\n",
       "      <td>ggtggtggtgctcgagtgcggccgcAGCATCTGCTGGAACACCGG</td>\n",
       "      <td>45</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0        gene       rv  Msm_VI Msm_Ess_TnSeq Msm_Ess_CRISPR  \\\n",
       "0           0  MSMEG_0250  Rv0206c  -7.702     Essential      Essential   \n",
       "1           1  MSMEG_0311   Rv0225  -0.060     Essential      Essential   \n",
       "2           2  MSMEG_0317  Rv0227c  -9.654     Essential      Essential   \n",
       "3           3  MSMEG_0359  Rv0236c  -0.614     Essential      Essential   \n",
       "4           4  MSMEG_0384   Rv0334  -9.876     Essential      Essential   \n",
       "\n",
       "   Mtb_VI Mtb_Ess_TnSeq Mtb_Ess_CRISPR  \\\n",
       "0 -11.013     Essential      Essential   \n",
       "1  -8.756     Essential      Essential   \n",
       "2 -13.167     Essential      Essential   \n",
       "3  -6.080     Essential      Essential   \n",
       "4  -2.498     Essential      Essential   \n",
       "\n",
       "                                             Msm_seq  ... Mtb_Aliphatic_Index  \\\n",
       "0  MFAWWGRTVYQFRYIVIGVMVALCLGGGVYGISLGNHVTQSGFYDE...  ...          102.658898   \n",
       "1  MSARPESAPHVRRVLLLCWRDTGHPQGGGSEAYVQRIGAYLAGRGV...  ...          105.598958   \n",
       "2  MNRAVALRIAACGLLGLGAALLIAALLLTTYTKGKIAKIPLDIDTS...  ...           76.294537   \n",
       "3  MVAAATLVLTFAQSPGQISPDTKLDLTANPLRFLARAFNLWNSDLP...  ...          100.614286   \n",
       "4  MRGIILAGGSGTRLHPLTIGVSKQLLPVYDKPLVYYPLSTLIMAGI...  ...           99.930556   \n",
       "\n",
       "  PI_difference Aliphatic_Index_difference  \\\n",
       "0      3.843500                   8.927408   \n",
       "1      0.352306                   6.928882   \n",
       "2      0.287468                   9.785865   \n",
       "3      3.159969                   2.127723   \n",
       "4      0.132063                   1.388889   \n",
       "\n",
       "                Msm Upstream Recombineering Sequence  \\\n",
       "0  GCAACGCCGTTCGCAATGCGGTCAACAGCGCGGTGCACGGCGGCGC...   \n",
       "1  AAGCCGTCGCGCAACTGCGCACCGAGATGCCTGATGTGCACCTGGA...   \n",
       "2  TCACCGCGCGCGCCGAGGTGTGGGGTGTGCCGGGCGAACCGGACGA...   \n",
       "3  ACATGCCGTACCGCGTCGGACTGATCGGCGGACTGGCACTCCTGCC...   \n",
       "4  GCTTCGAACACGTCAGTGGTGGGGCAATATTCGCGTACTGGGTGGC...   \n",
       "\n",
       "              Msm Downstream Recombineering Sequence  \\\n",
       "0  GGCCCGGCTACTCCCGTTCGGGAAGGTCGTCGGACTCTGCGGCGAT...   \n",
       "1  CGCCGCTCAGAACGCCTCGGCGGCTGCCTTCGCGGTGCGGGCCAGC...   \n",
       "2  GAACTGCGGTCCTGCTGCGGTCATGCTCGGTGCCGGCCTATGCGCT...   \n",
       "3  CCAATCCTTGCCACCTCAACTGGTTGACAGCGAGCCGGTGCAGTAC...   \n",
       "4  CGTGCCTAGCGCAAGGGCATACTGCTGTAACACCTGGCATGCCCAC...   \n",
       "\n",
       "                         Msm Fwd Upstream primer  \\\n",
       "0     tttaagaaggagatatacatatgGCAACGCCGTTCGCAATGC   \n",
       "1     tttaagaaggagatatacatatgAAGCCGTCGCGCAACTGCG   \n",
       "2     tttaagaaggagatatacatatgTCACCGCGCGCGCCGAGGT   \n",
       "3     tttaagaaggagatatacatatgACATGCCGTACCGCGTCGG   \n",
       "4  tttaagaaggagatatacatatgGCTTCGAACACGTCAGTGGTGG   \n",
       "\n",
       "                       Msm Rvs Upstream primer  \\\n",
       "0        accgctgcccgagccaccCAGCCTGCCTTCGCGGCGC   \n",
       "1        accgctgcccgagccaccGACCAGGCCGCTGAGGTAT   \n",
       "2        accgctgcccgagccaccGATCGGTCGGTCCGGTGGC   \n",
       "3        accgctgcccgagccaccACCGGCGTGAGCGGATGCT   \n",
       "4  accgctgcccgagccaccCTCTCGATCCAGAAGTTGGAGCAGG   \n",
       "\n",
       "                  Msm Fwd Downstream primer  \\\n",
       "0     atctcgagtctagaagtaGGCCCGGCTACTCCCGTTC   \n",
       "1     atctcgagtctagaagtaCGCCGCTCAGAACGCCTCG   \n",
       "2     atctcgagtctagaagtaGAACTGCGGTCCTGCTGCG   \n",
       "3  atctcgagtctagaagtaCCAATCCTTGCCACCTCAACTG   \n",
       "4     atctcgagtctagaagtaCGTGCCTAGCGCAAGGGCA   \n",
       "\n",
       "                        Msm Rvs Downstream primer  \\\n",
       "0    ggtggtggtgctcgagtgcggccgcCGTGCGCGCGACGTTCCTG   \n",
       "1    ggtggtggtgctcgagtgcggccgcCCGCGGCGCTCGACGCCAT   \n",
       "2  ggtggtggtgctcgagtgcggccgcCATCCAGAACACCAGCGGAGC   \n",
       "3    ggtggtggtgctcgagtgcggccgcCGCGCTGTACGGCGACGAC   \n",
       "4   ggtggtggtgctcgagtgcggccgcAGCATCTGCTGGAACACCGG   \n",
       "\n",
       "   Msm max recombineering primer length  \n",
       "0                                    44  \n",
       "1                                    44  \n",
       "2                                    46  \n",
       "3                                    44  \n",
       "4                                    45  \n",
       "\n",
       "[5 rows x 30 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_df.head()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
