{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "57d265d3-5115-4462-b1a1-52dffc8e682e",
   "metadata": {},
   "source": [
    "### Importing packages and functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e81c6220-d43a-4d09-9a66-a66ed6b16fa4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns \n",
    "import matplotlib.pyplot as plt\n",
    "import os\n",
    "import peptides as pep\n",
    "from Bio.Seq import Seq \n",
    "from Bio.SeqUtils import MeltingTemp as mt\n",
    "from Bio import SeqIO\n",
    "from Bio.SeqUtils import GC\n",
    "\n",
    "#Load protein fasta files with Biopython \n",
    "\n",
    "def read_fasta(fasta_file):\n",
    "    content = {}\n",
    "    fas = open(fasta_file,'r')\n",
    "    seq = ''\n",
    "    last_entry = ''\n",
    "    while True:\n",
    "        line = fas.readline()\n",
    "        if len(line)==0:\n",
    "            content[last_entry]=seq\n",
    "            break\n",
    "        if '>' in line:\n",
    "            if last_entry=='':\n",
    "                line = line.rstrip()\n",
    "                last_entry=line\n",
    "            else:\n",
    "                line = line.rstrip()\n",
    "                content[last_entry]=seq\n",
    "                last_entry=line\n",
    "                seq=''\n",
    "        else:\n",
    "            line = line.rstrip()\n",
    "            seq += line\n",
    "    return content\n",
    "\n",
    "def read_gff(gff_file, retrieve_keys = ['Locus','Name']):\n",
    "    import pandas as pd\n",
    "    columns=['Seqid','Source','type','Start',\n",
    "             'End','Score','Strand','Phase'] + retrieve_keys\n",
    "    df_raw = []\n",
    "    with open(gff_file, 'r') as f:\n",
    "        while True:\n",
    "            line = f.readline().rstrip()\n",
    "            if line == '':\n",
    "                f.close()\n",
    "                break\n",
    "            else:\n",
    "                line = line.split('\\t')\n",
    "                info = line[:-1]\n",
    "                info_dict={}\n",
    "                for x in line[-1].split(';'):\n",
    "                    if x.count('=') == 1:\n",
    "                        k,v = x.split('=')\n",
    "                        info_dict[k] = v\n",
    "                info += [info_dict[k] for k in retrieve_keys]\n",
    "                df_raw.append(info)\n",
    "    df = pd.DataFrame(df_raw,columns=columns)\n",
    "    df[['Start','End']] = df[['Start','End']].astype(int)\n",
    "    return df.sort_values(by='Start').reset_index(drop=True)\n",
    "\n",
    "\n",
    "def determine_primers(gene_sequence, gibson_handle, direction):\n",
    "        \n",
    "    #Compute reverse complement for reverse primers and check for a stop codon        \n",
    "    if direction == 'rvs':\n",
    "        protein = gene_sequence.translate()\n",
    "        test_stop = protein.endswith('*')\n",
    "        if test_stop:\n",
    "            gene_sequence = gene_sequence[0:len(gene_sequence)-3]\n",
    "        gene_sequence = gene_sequence.reverse_complement()\n",
    "    \n",
    "    #Remove the start codon if fwd seq\n",
    "    if direction == 'fwd':\n",
    "        gene_sequence = gene_sequence[3:]    \n",
    "        \n",
    "    #Grabbing the first nucleotide to initialize a seq object\n",
    "    pcr_handle = gene_sequence[0]\n",
    "    \n",
    "    #Extend PCR handle based on gene sequence until a melting temperature of 58 reached\n",
    "    i = 1\n",
    "    while mt.Tm_NN(pcr_handle, nn_table=mt.DNA_NN4)<58:\n",
    "        pcr_handle = pcr_handle + gene_sequence[i]\n",
    "        i = i+1\n",
    "        \n",
    "    #Set the minumum length of the PCR handle to 19 nucleotides\n",
    "    if i < 19:\n",
    "        pcr_handle = pcr_handle + gene_sequence[i:19]\n",
    "        i = 19\n",
    "        \n",
    "    #Returns primer sequence and melting temp\n",
    "    primer = gibson_handle + pcr_handle\n",
    "    return(str(primer))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8dd9d143-66dd-4db7-b5f7-b77d9f376c7f",
   "metadata": {},
   "source": [
    "### Loading reference data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "3019dc3b-7323-4df2-9a98-fdf3cb6570e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "path = './Important_References/'\n",
    "union_ortho_table_filtered = pd.read_csv(path+'union_ortho_table_filtered.csv')\n",
    "\n",
    "#Loading Msm gene sequences with locus ID \n",
    "msm_gene_seqs = {k[1:].split('|')[0]:v for k,v in read_fasta(path+'Mycobacterium_smegmatis_MC2-155_genes_v4.fasta').items()}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "73bcf904-ff38-46c6-97e1-518e964a7605",
   "metadata": {},
   "source": [
    "### Design primers for all substrates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "d8f6cb56-b53a-4d67-8345-c7b6edabcdcd",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Set the gibson handles based on the sequence of the digested plasmid pHW76 \n",
    "Fwd_gibson_handle = Seq('TAATTAATATCAGGAGGTATACATatg').upper()\n",
    "Rvs_gibson_handle = Seq('CagcGctGCCCGAGCCACCAGTACT').upper()\n",
    "\n",
    "union_ortho_table_full_length = union_ortho_table_filtered.copy()\n",
    "union_ortho_table_full_length['Msm_Fwd_primer'] = [determine_primers(Seq(msm_gene_seqs[x]), Fwd_gibson_handle, 'fwd') for x in union_ortho_table_full_length['gene'].values]\n",
    "union_ortho_table_full_length['Msm_Rvs_primer'] = [determine_primers(Seq(msm_gene_seqs[x]), Rvs_gibson_handle, 'rvs') for x in union_ortho_table_full_length['gene'].values]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "fa583640-89b6-43f3-b6f6-60ca99c7a23f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Save to csv\n",
    "union_ortho_table_full_length.to_csv(path+'union_ortho_table_full_length.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "dfc4be96-5445-4879-b74c-d75d0217f2c2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>gene</th>\n",
       "      <th>rv</th>\n",
       "      <th>Msm_VI</th>\n",
       "      <th>Msm_Ess_TnSeq</th>\n",
       "      <th>Msm_Ess_CRISPR</th>\n",
       "      <th>Mtb_VI</th>\n",
       "      <th>Mtb_Ess_TnSeq</th>\n",
       "      <th>Mtb_Ess_CRISPR</th>\n",
       "      <th>Msm_seq</th>\n",
       "      <th>...</th>\n",
       "      <th>Msm_pro_length</th>\n",
       "      <th>Mtb_pro_length</th>\n",
       "      <th>Msm_PI</th>\n",
       "      <th>Mtb_PI</th>\n",
       "      <th>Msm_Aliphatic_Index</th>\n",
       "      <th>Mtb_Aliphatic_Index</th>\n",
       "      <th>PI_difference</th>\n",
       "      <th>Aliphatic_Index_difference</th>\n",
       "      <th>Msm_Fwd_primer</th>\n",
       "      <th>Msm_Rvs_primer</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>MSMEG_0250</td>\n",
       "      <td>Rv0206c</td>\n",
       "      <td>-7.702</td>\n",
       "      <td>Essential</td>\n",
       "      <td>Essential</td>\n",
       "      <td>-11.013</td>\n",
       "      <td>Essential</td>\n",
       "      <td>Essential</td>\n",
       "      <td>MFAWWGRTVYQFRYIVIGVMVALCLGGGVYGISLGNHVTQSGFYDE...</td>\n",
       "      <td>...</td>\n",
       "      <td>1013</td>\n",
       "      <td>944</td>\n",
       "      <td>6.063635</td>\n",
       "      <td>9.907134</td>\n",
       "      <td>93.731491</td>\n",
       "      <td>102.658898</td>\n",
       "      <td>3.843500</td>\n",
       "      <td>8.927408</td>\n",
       "      <td>TAATTAATATCAGGAGGTATACATATGttcgcctggtggggtcgga</td>\n",
       "      <td>CAGCGCTGCCCGAGCCACCAGTACTcagcctgccttcgcggcgc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>MSMEG_0311</td>\n",
       "      <td>Rv0225</td>\n",
       "      <td>-0.060</td>\n",
       "      <td>Essential</td>\n",
       "      <td>Essential</td>\n",
       "      <td>-8.756</td>\n",
       "      <td>Essential</td>\n",
       "      <td>Essential</td>\n",
       "      <td>MSARPESAPHVRRVLLLCWRDTGHPQGGGSEAYVQRIGAYLAGRGV...</td>\n",
       "      <td>...</td>\n",
       "      <td>391</td>\n",
       "      <td>384</td>\n",
       "      <td>9.791186</td>\n",
       "      <td>10.143492</td>\n",
       "      <td>98.670077</td>\n",
       "      <td>105.598958</td>\n",
       "      <td>0.352306</td>\n",
       "      <td>6.928882</td>\n",
       "      <td>TAATTAATATCAGGAGGTATACATATGtctgcccggcccgagtctg</td>\n",
       "      <td>CAGCGCTGCCCGAGCCACCAGTACTgaccaggccgctgaggtat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>MSMEG_0317</td>\n",
       "      <td>Rv0227c</td>\n",
       "      <td>-9.654</td>\n",
       "      <td>Essential</td>\n",
       "      <td>Essential</td>\n",
       "      <td>-13.167</td>\n",
       "      <td>Essential</td>\n",
       "      <td>Essential</td>\n",
       "      <td>MNRAVALRIAACGLLGLGAALLIAALLLTTYTKGKIAKIPLDIDTS...</td>\n",
       "      <td>...</td>\n",
       "      <td>398</td>\n",
       "      <td>421</td>\n",
       "      <td>4.292958</td>\n",
       "      <td>4.580426</td>\n",
       "      <td>86.080402</td>\n",
       "      <td>76.294537</td>\n",
       "      <td>0.287468</td>\n",
       "      <td>9.785865</td>\n",
       "      <td>TAATTAATATCAGGAGGTATACATATGaaccgcgctgtggcgctgc</td>\n",
       "      <td>CAGCGCTGCCCGAGCCACCAGTACTgatcggtcggtccggtggc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>MSMEG_0359</td>\n",
       "      <td>Rv0236c</td>\n",
       "      <td>-0.614</td>\n",
       "      <td>Essential</td>\n",
       "      <td>Essential</td>\n",
       "      <td>-6.080</td>\n",
       "      <td>Essential</td>\n",
       "      <td>Essential</td>\n",
       "      <td>MVAAATLVLTFAQSPGQISPDTKLDLTANPLRFLARAFNLWNSDLP...</td>\n",
       "      <td>...</td>\n",
       "      <td>1414</td>\n",
       "      <td>1400</td>\n",
       "      <td>6.699455</td>\n",
       "      <td>9.859423</td>\n",
       "      <td>98.486563</td>\n",
       "      <td>100.614286</td>\n",
       "      <td>3.159969</td>\n",
       "      <td>2.127723</td>\n",
       "      <td>TAATTAATATCAGGAGGTATACATATGgtcgccgcggcgacactcg</td>\n",
       "      <td>CAGCGCTGCCCGAGCCACCAGTACTaccggcgtgagcggatgct</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>MSMEG_0384</td>\n",
       "      <td>Rv0334</td>\n",
       "      <td>-9.876</td>\n",
       "      <td>Essential</td>\n",
       "      <td>Essential</td>\n",
       "      <td>-2.498</td>\n",
       "      <td>Essential</td>\n",
       "      <td>Essential</td>\n",
       "      <td>MRGIILAGGSGTRLHPLTIGVSKQLLPVYDKPLVYYPLSTLIMAGI...</td>\n",
       "      <td>...</td>\n",
       "      <td>288</td>\n",
       "      <td>288</td>\n",
       "      <td>5.004825</td>\n",
       "      <td>5.136888</td>\n",
       "      <td>101.319444</td>\n",
       "      <td>99.930556</td>\n",
       "      <td>0.132063</td>\n",
       "      <td>1.388889</td>\n",
       "      <td>TAATTAATATCAGGAGGTATACATATGcgcggcatcatcctcgccg</td>\n",
       "      <td>CAGCGCTGCCCGAGCCACCAGTACTctctcgatccagaagttggag...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0        gene       rv  Msm_VI Msm_Ess_TnSeq Msm_Ess_CRISPR  \\\n",
       "0           0  MSMEG_0250  Rv0206c  -7.702     Essential      Essential   \n",
       "1           1  MSMEG_0311   Rv0225  -0.060     Essential      Essential   \n",
       "2           2  MSMEG_0317  Rv0227c  -9.654     Essential      Essential   \n",
       "3           3  MSMEG_0359  Rv0236c  -0.614     Essential      Essential   \n",
       "4           4  MSMEG_0384   Rv0334  -9.876     Essential      Essential   \n",
       "\n",
       "   Mtb_VI Mtb_Ess_TnSeq Mtb_Ess_CRISPR  \\\n",
       "0 -11.013     Essential      Essential   \n",
       "1  -8.756     Essential      Essential   \n",
       "2 -13.167     Essential      Essential   \n",
       "3  -6.080     Essential      Essential   \n",
       "4  -2.498     Essential      Essential   \n",
       "\n",
       "                                             Msm_seq  ... Msm_pro_length  \\\n",
       "0  MFAWWGRTVYQFRYIVIGVMVALCLGGGVYGISLGNHVTQSGFYDE...  ...           1013   \n",
       "1  MSARPESAPHVRRVLLLCWRDTGHPQGGGSEAYVQRIGAYLAGRGV...  ...            391   \n",
       "2  MNRAVALRIAACGLLGLGAALLIAALLLTTYTKGKIAKIPLDIDTS...  ...            398   \n",
       "3  MVAAATLVLTFAQSPGQISPDTKLDLTANPLRFLARAFNLWNSDLP...  ...           1414   \n",
       "4  MRGIILAGGSGTRLHPLTIGVSKQLLPVYDKPLVYYPLSTLIMAGI...  ...            288   \n",
       "\n",
       "  Mtb_pro_length    Msm_PI     Mtb_PI Msm_Aliphatic_Index  \\\n",
       "0            944  6.063635   9.907134           93.731491   \n",
       "1            384  9.791186  10.143492           98.670077   \n",
       "2            421  4.292958   4.580426           86.080402   \n",
       "3           1400  6.699455   9.859423           98.486563   \n",
       "4            288  5.004825   5.136888          101.319444   \n",
       "\n",
       "   Mtb_Aliphatic_Index  PI_difference  Aliphatic_Index_difference  \\\n",
       "0           102.658898       3.843500                    8.927408   \n",
       "1           105.598958       0.352306                    6.928882   \n",
       "2            76.294537       0.287468                    9.785865   \n",
       "3           100.614286       3.159969                    2.127723   \n",
       "4            99.930556       0.132063                    1.388889   \n",
       "\n",
       "                                   Msm_Fwd_primer  \\\n",
       "0  TAATTAATATCAGGAGGTATACATATGttcgcctggtggggtcgga   \n",
       "1  TAATTAATATCAGGAGGTATACATATGtctgcccggcccgagtctg   \n",
       "2  TAATTAATATCAGGAGGTATACATATGaaccgcgctgtggcgctgc   \n",
       "3  TAATTAATATCAGGAGGTATACATATGgtcgccgcggcgacactcg   \n",
       "4  TAATTAATATCAGGAGGTATACATATGcgcggcatcatcctcgccg   \n",
       "\n",
       "                                      Msm_Rvs_primer  \n",
       "0       CAGCGCTGCCCGAGCCACCAGTACTcagcctgccttcgcggcgc  \n",
       "1       CAGCGCTGCCCGAGCCACCAGTACTgaccaggccgctgaggtat  \n",
       "2       CAGCGCTGCCCGAGCCACCAGTACTgatcggtcggtccggtggc  \n",
       "3       CAGCGCTGCCCGAGCCACCAGTACTaccggcgtgagcggatgct  \n",
       "4  CAGCGCTGCCCGAGCCACCAGTACTctctcgatccagaagttggag...  \n",
       "\n",
       "[5 rows x 25 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "union_ortho_table_full_length.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "58e5b08d-fe2e-4c8e-80dd-fe1b7888bec5",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
